plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java

```diff
@@ -0,0 +1,157 @@
+		QueryAnnotationsEvent.AnnotationFilter filter = new QueryAnnotationsEvent.AnnotationFilter() {
+
+			@Override
+			public boolean accept(Annotation annotation) {
+				return annotation instanceof DiagnosticAnnotation;
+			}
+
+		};
+		QueryAnnotationsEvent.QueryCallback annotationCallback = new QueryAnnotationsEvent.QueryCallback() {
+
+			@Override
+			public void respond(Map<Annotation, Position> annotations) {
+				List<DiagnosticDTO> diagnostics = new ArrayList<>();
+				// iteration with range never returns anything; need to filter ourselves.
+				// https://github.com/eclipse/che/issues/4338
+				annotations.entrySet().stream().filter((e)->e.getValue().overlapsWith(range.getStartOffset(), range.getLength())).forEach((e) -> diagnostics.add(((DiagnosticAnnotation) e.getKey()).getDiagnostic()));
+				CodeActionContextDTO context = CodeActionContextDTOImpl.make().withDiagnostics(diagnostics);
+				CodeActionParamsDTO params = CodeActionParamsDTOImpl.make().withContext(context);
+
+				params.setTextDocument(
+								TextDocumentIdentifierDTOImpl.make().withUri(document.getFile().getLocation().toString()));
+
+				TextPosition start = document.getPositionFromIndex(range.getStartOffset());
+				TextPosition end = document.getPositionFromIndex(range.getStartOffset() + range.getLength());
+				PositionDTO rangeStart = PositionDTOImpl.make().withLine(start.getLine());
+				rangeStart.setCharacter(start.getCharacter());
+				PositionDTO rangeEnd = PositionDTOImpl.make().withLine(end.getLine());
+				rangeEnd.setCharacter(end.getCharacter());
+				RangeDTO rangeParam = RangeDTOImpl.make().withStart(rangeStart);
+				rangeParam.setEnd(rangeEnd);
+				params.setRange(rangeParam);
+
+				Promise<List<CommandDTO>> codeAction = textDocumentService.codeAction(params);
+				List<CompletionProposal> proposals = new ArrayList<>();
+				codeAction.then((commands) -> {
+					for (CommandDTO command : commands) {
+						Action action = actionManager.getAction(command.getCommand());
+						if (action != null) {
+
+							proposals.add(new CompletionProposal() {
+
+								@Override
+								public void getAdditionalProposalInfo(AsyncCallback<Widget> callback) {
+								}
+
+								@Override
+								public String getDisplayString() {
+									return command.getTitle();
+								}
+
+								@Override
+								public Icon getIcon() {
+									return null;
+								}
+
+								@Override
+								public void getCompletion(CompletionCallback callback) {
+									callback.onCompletion(new Completion() {
+
+										@Override
+										public LinearRange getSelection(Document document) {
+											return null;
+										}
+
+										@Override
+										public void apply(Document document) {
+											QuickassistActionEvent evt = new QuickassistActionEvent(new Presentation(), actionManager, perspectiveManager, command.getArguments());
+											action.actionPerformed(evt);
+										}
```

Comment:
I understand your point of view however I can't tell that I can agree with your arguments, I've done small refactoring so could you please tell me if you still think that your variant is more preferable to something like this:
```java
public class LanguageServerQuickAssistProcessor implements QuickAssistProcessor {
	private TextDocumentServiceClient textDocumentService;
	private ActionManager actionManager;
	private PerspectiveManager perspectiveManager;

	@Inject
	public LanguageServerQuickAssistProcessor(TextDocumentServiceClient textDocumentService, ActionManager actionManager, PerspectiveManager perspectiveManager) {
		this.textDocumentService = textDocumentService;
		this.actionManager = actionManager;
		this.perspectiveManager= perspectiveManager;
	}

	@Override
	public void computeQuickAssistProposals(QuickAssistInvocationContext invocationContext, CodeAssistCallback callback) {
		LinearRange range = invocationContext.getTextEditor().getSelectedLinearRange();
		Document document = invocationContext.getTextEditor().getDocument();
        QueryAnnotationsEvent.AnnotationFilter filter = annotation -> annotation instanceof DiagnosticAnnotation;

        QueryAnnotationsEvent.QueryCallback annotationCallback = annotations -> {
            // iteration with range never returns anything; need to filter ourselves.
            // https://github.com/eclipse/che/issues/4338
            List<DiagnosticDTO> diagnostics = getAnnotations(annotations, range.getStartOffset(), range.getLength());
            CodeActionContextDTO context = CodeActionContextDTOImpl.make().withDiagnostics(diagnostics);
            CodeActionParamsDTO params = CodeActionParamsDTOImpl.make().withContext(context);

            params.setTextDocument(TextDocumentIdentifierDTOImpl.make().withUri(document.getFile().getLocation().toString()));

            TextPosition start = document.getPositionFromIndex(range.getStartOffset());
            TextPosition end = document.getPositionFromIndex(range.getStartOffset() + range.getLength());
            PositionDTO rangeStart = PositionDTOImpl.make().withLine(start.getLine());
            rangeStart.setCharacter(start.getCharacter());
            PositionDTO rangeEnd = PositionDTOImpl.make().withLine(end.getLine());
            rangeEnd.setCharacter(end.getCharacter());
            RangeDTO rangeParam = RangeDTOImpl.make().withStart(rangeStart);
            rangeParam.setEnd(rangeEnd);
            params.setRange(rangeParam);

            textDocumentService.codeAction(params).then(processCommandProposals(callback));
        };

        QueryAnnotationsEvent event = new QueryAnnotationsEvent.Builder().withFilter(filter).withCallback(annotationCallback).build();
        document.getDocumentHandle().getDocEventBus().fireEvent(event);
    }

    private Operation<List<CommandDTO>> processCommandProposals(CodeAssistCallback callback) {
        return commands -> {
            List<CompletionProposal> proposals = commands.stream()
                                                         .map(command -> new CommandActionPair(command, actionManager))
                                                         .filter(pair -> pair.action != null)
                                                         .map(QuickAssistCompletionProposal::new)
                                                         .collect(toList());
            callback.proposalComputed(proposals);
        };
    }

    private List<DiagnosticDTO> getAnnotations(Map<Annotation, Position> annotations, int offset, int length) {
        return annotations.entrySet()
                          .stream()
                          .filter(it -> it.getValue().overlapsWith(offset, length))
                          .map(Map.Entry::getKey)
                          .map(it -> (DiagnosticAnnotation)it)
                          .map(DiagnosticAnnotation::getDiagnostic)
                          .collect(toList());
    }

    private class QuickAssistCompletionProposal implements CompletionProposal {
        private final CommandActionPair pair;

        private QuickAssistCompletionProposal(CommandActionPair pair) {
            this.pair = pair;
        }

        @Override
        public void getAdditionalProposalInfo(AsyncCallback<Widget> callback1) {
        }

        @Override
        public String getDisplayString() {
            return pair.command.getTitle();
        }

        @Override
        public Icon getIcon() {
            return null;
        }

        @Override
        public void getCompletion(CompletionCallback completionCallback) {
            completionCallback.onCompletion(new Completion() {

                @Override
                public LinearRange getSelection(Document document) {
                    return null;
                }

                @Override
                public void apply(Document document) {
                    Presentation presentation = new Presentation();
                    List<Object> arguments = pair.command.getArguments();
                    pair.action.actionPerformed(new QuickassistActionEvent(presentation, actionManager, perspectiveManager, arguments));
                }
            });
        }
    }

    private class CommandActionPair {
        final private Command command;
        final private Action  action;

        private CommandActionPair(Command command, ActionManager actionManager) {
            this.command = command;
            this.action = actionManager.getAction(command.getCommand());
        }
    }
}
```
What you do is a hard work, complex business logics and architecture I respect that, however my intentions here are to accept the code that represents complex business logic as simple as possible so any committer could contribute, so please don't take my comments negatively.